### ISSUE TITLE
Couldn't aquire lock, is another instance running?

### ISSUE THREAD

**Issue by peters** [Squirrel/Squirrel.Windows#138]
After updating to 6.0.1 many of our logs are flooded with these exceptions. Has something changed? Can't spot anything wrong with the code below either. `CheckForUpdates` method is only called from a single place in the application. The update still works though after a couple of retries when it's able to aquire the lock.

/cc @paulcbetts 

```csharp
YggdrasilSetup.CheckForUpdates((version, updateAbsPath) =>
{
                    Info("Successfully updated yggdrasil to new version: {0}. Assembly location is now: {1}", version, updateAbsPath);
                    _updated = true;
                    YggdrasilSetup.OnAppStart(updateAbsPath);
                    YggdrasilAssembly.RestartApplication(updateAbsPath, RestartApplicationReason.Update);
 }).Wait(CancellationToken);
```

```csharp
internal static async Task<bool> CheckForUpdates(Action<Version, string> updateSuccess = null)
        {
            // ReSharper disable once JoinDeclarationAndInitializer
            string updateUrl;

#if INSTALLER
            updateUrl = @"C:\installers\yggdrasil\{0}".Fmt(IntPtr.Size == 4 ? "x86" : "x64");
#elif DEPLOY_PRODUCTION
            updateUrl = "http://software.odinkapital.no/apps/yggdrasil/production/updates";
#elif DEPLOY_STAGING
            updateUrl = "http://software.odinkapital.no/apps/yggdrasil/staging/updates";
#else
            return false;
#endif

#pragma warning disable 162
            if (updateSuccess == null) updateSuccess = (a, b) => { };

            string updateAbsPath = null;
            ReleaseEntry releaseEntry = null;

            using (var updateManager = new UpdateManager(updateUrl, YggdrasilAssembly.AssemblyName, FrameworkVersion.Net45))
            {
                var updateInfo = await updateManager.CheckForUpdate(true);
                if (updateInfo != null && updateInfo.ReleasesToApply.Count > 0)
                {
                    releaseEntry = updateInfo.ReleasesToApply.OrderByDescending(x => x.Version).First();
                    await updateManager.DownloadReleases(updateInfo.ReleasesToApply, i => Logger.Info("DownloadReleases progress: {0}", i));
                    updateAbsPath = await updateManager.ApplyReleases(updateInfo, i => Logger.Info("ApplyReleases progress: {0}", i));
                    updateAbsPath = Path.Combine(updateAbsPath, "yggdrasil.exe");
                }
            }

            if (updateAbsPath == null)
            {
                return false;
            }

            updateSuccess(releaseEntry.Version, updateAbsPath);

            return true;
#pragma warning restore 162
        }
```

```csharp
11/25/2014 09:55:33 Process id: 15948, Version: 1.9.10.0, Thread:  - [CheckForUpdatesJob]: Forever loop EXCEPTION OCCURRED:System.AggregateException: One or more errors occurred. ---> System.Exception: Couldn't acquire lock, is another instance running
   at Squirrel.SingleGlobalInstance..ctor(String key, TimeSpan timeOut)
   at Squirrel.UpdateManager.<acquireUpdateLock>b__1e()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Squirrel.UpdateManager.<CheckForUpdate>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at ...()
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait(CancellationToken cancellationToken)
   at odinoffice.yggdrasil.Src.Jobs.CheckForUpdatesJob.<<Run>b__0>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at ...()
---> (Inner Exception #0) System.Exception: Couldn't acquire lock, is another instance running
   at Squirrel.SingleGlobalInstance..ctor(String key, TimeSpan timeOut)
   at Squirrel.UpdateManager.<acquireUpdateLock>b__1e()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Squirrel.UpdateManager.<CheckForUpdate>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at ...()<---
    at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait(CancellationToken cancellationToken)
   at odinoffice.yggdrasil.Src.Jobs.CheckForUpdatesJob.<<Run>b__0>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at ...()
```

**Comment by dealproc** [comment]
In certain scenarios, I'm experiencing this too with a demo app. have you gotten any resolution to this @peters

**Comment by theofanis** [comment]
I also have this problem. It occurs after having checked for update at least once. All subsequent calls to ```CheckForUpdate``` throw this exception. This is how I check for updates:

```
public static async Task<UpdateInfo> CheckForUpdateAsync()
        {
            var appName = GetAppName();
            var updateLocation = GetUpdateLocation();

            try
            {
                _log.Debug("Checking for update. Called " + Utilities.GetCallerName() + " with UpdateLocation = \"" + updateLocation + "\" and AppName = " + appName + "\".");

                using (var mgr = new Squirrel.UpdateManager(updateLocation, appName, Squirrel.FrameworkVersion.Net45))
                {
                    return await mgr.CheckForUpdate();
                }
            }
            catch(Exception e)
            {
                _log.Error(Utilities.GetCallerName() + " error while checking for update with UpdateLocation = \"" + updateLocation + "\" and AppName = " + appName + ", thrown " + e.GetType().ToString() + " with message \"" + e.Message + "\".", e);
            }

            return null;
        }
```

**Comment by theofanis** [comment]
I still can't figure this one out. 

@paulcbetts your insight please?

**Comment by paulcbetts** [comment]
Fixed in #189
